MinimedKit/PumpManager/MinimedDoseProgressEstimator.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  MinimedDoseProgressEstimator.swift
+//  MinimedKit
+//
+//  Created by Pete Schwamb on 3/14/19.
+//  Copyright © 2019 Pete Schwamb. All rights reserved.
+//
+
+import Foundation
+import LoopKit
+
+class MinimedDoseProgressEstimator: DoseProgressEstimator {
+
+    public let dose: DoseEntry
+
+    public let pumpModel: PumpModel
+
+    private var observers = WeakSet<DoseProgressObserver>()
+
+    private var lock = os_unfair_lock()
+
+    private var timer: Timer?
+
+    var progress: DoseProgress {
+        let elapsed = -dose.startDate.timeIntervalSinceNow
+        let duration = dose.endDate.timeIntervalSince(dose.startDate)
+        let timeProgress = min(elapsed / duration, 1)
+
+        let updateResolution: Double
+        let unroundedVolume: Double
+
+        if pumpModel.variableDeliveryRate {
+            if dose.units < 1 {
+                updateResolution = 40 // Resolution = 0.025
+                unroundedVolume = timeProgress * dose.units
+            } else {
+                var remainingUnits = dose.units
+                var baseDuration: TimeInterval = 0
+                var overlay1Duration: TimeInterval = 0
+                var overlay2Duration: TimeInterval = 0
+                let baseDeliveryRate = 1.5 / TimeInterval(minutes: 1)
+
+                baseDuration = min(duration, remainingUnits / baseDeliveryRate)
```

Comment:
Looks like line #s changed to 37-46.

Reply:
The base delivery rate is 1.5 U/m, which is a 0.05 pulse every 2s.  If, at this rate, it cannot deliver the entire dose within 5 minutes, it starts double pulsing (0.1U), but keeping the same periodicity (every 2s). It double pulses for just long enough that the remainder will finish out in 5 minutes at the lower 1.5U/m, single pulse rate.  If the dose is large enough that it can't deliver in 5 minutes even by double pulsing, it will start triple pulsing.  I didn't know this was precisely how it did it until last week, when I had to figure it out. :)

Reply:
Thanks for the clarification. So, `duration` is capped at a maximum of 5 minutes elsewhere?

Reply:
Duration is calculated by PumpModel.bolusDeliveryTime, and for x23 pumps, it will be a max of 5 minutes

Reply:
:+1: